Yoast/wordpress-seo
# P3 149 monitor taxomony base changes

## Context


* Currently the taxonomy base might have changed without our detection. For instance because it was directly changed in the database, or dynamically overwritten by a plugin, or updated through WP CLI. This means permalinks stored with indexables will be incorrect.

## Summary


This PR can be summarized in the following changelog entry:

* We store the category base url and the tag base url in the wpseo options. Then when we create a new indexable, we need to validate the taxonomy base matches the one we stored for that taxonomy. If any difference occurs, we can clear all permalinks, store the new permalink structure and ask for a reindex.

## Relevant technical choices:

* Added two new options to keep track of the previous base urls of category and tag

## Test instructions

This PR can be tested by following these steps:

* Checkout this branch, build, reactivate the plugin 
* Make sure to have 25 or more categories and 25 or more tags.
*  Set your category base url and tag base url to empty from the permalink settings
* Run the upgrade routine (by using the test helper)
* Run the indexing from Yoast SEO -> tools if available
* Prepare your database program to view and change data ( i highly recommend Sequel Pro )
* Check the wpseo options and see that category_base_url and tag_base_url look empty like this:
`category_base_url";s:0:"";s:12:"tag_base_url";s:0:"`
1. The category test
* Go to the wpseo admin dashboard and notice there is no notification stating you should reindex
* Run `./wp.sh cron event run wpseo_permalink_structure_check` within your docker environment
* Refresh your wpseo admin dashboard and notice that the notification is still not there
* Go to your database explorer and find the option named category_base in the wp_options table and change its value and save
* Refresh your wpseo admin dashboard and notice that the notification is still not there
* Run `./wp.sh cron event run wpseo_permalink_structure_check` within your docker environment
* Now see that there is a notification which looks exactly like this:
![Screenshot_2020-10-20 General - Yoast SEO ‹ Basic — WordPress](https://user-images.githubusercontent.com/15989132/96597978-aeda7180-12ee-11eb-8a31-d655206c19c4.png)
* Now change your category_base value to empty again, run `./wp.sh cron event run wpseo_permalink_structure_check` within your docker environment and reindex

2. The tag test
* Make sure your options are reset to empty
* Go to the wpseo admin dashboard and notice there is no notification stating you should reindex
* Run `./wp.sh cron event run wpseo_permalink_structure_check` within your docker environment
* Refresh your wpseo admin dashboard and notice that the notification is still not there
* Go to your database explorer and find the option named tag_base in the wp_options table and change its value and save
* Refresh your wpseo admin dashboard and notice that the notification is still not there
* Run `./wp.sh cron event run wpseo_permalink_structure_check` within your docker environment
* Now see that there is a notification which looks exactly like this:
![Screenshot_2020-10-20 General - Yoast SEO ‹ Basic — WordPress(1)](https://user-images.githubusercontent.com/15989132/96598307-0678dd00-12ef-11eb-9a15-93fb9b2c8937.png)
* Now change your tag_base value to empty again, run `./wp.sh cron event run wpseo_permalink_structure_check` within your docker environment and reindex

* Also try to test this on multisite (with 25+posts on that site) and see if it works the same there. In that case the CLI command will need to know the URL of the subsite you're reffering to, e.g.:
`./wp.sh cron event run wpseo_permalink_structure_check --url=site2.multisite.wordpress.test`
* Try to break this :D 



## UI changes

## Documentation

## Quality assurance


Fixes #
